vochain/account.go

```diff
@@ -175,108 +214,502 @@ func (v *State) GetAccount(address common.Address, isQuery bool) (*Account, erro
-	} else if err == nil {
-		if err := acc.Unmarshal(raw); err != nil {
+	}
+	if accountAddress == txSender {
+		acc.InfoURI = infoURI
+		acc.Balance -= SetAccountInfoCost
+		acc.Nonce++
+		accBytes, err := acc.Marshal()
+		if err != nil {
 			return err
 		}
+		v.Tx.Lock()
+		defer v.Tx.Unlock()
+		return v.Tx.DeepSet(accountAddress.Bytes(), accBytes, AccountsCfg)
+	}
+	sender, err := v.GetAccount(txSender, false)
+	if err != nil {
+		return err
+	}
+	if sender == nil {
+		return ErrAccountNotFound
+	}
+	sender.Balance -= SetAccountInfoCost
+	sender.Nonce++
+	senderBytes, err := sender.Marshal()
+	if err != nil {
+		return err
 	}
 	acc.InfoURI = infoURI
 	accBytes, err := acc.Marshal()
 	if err != nil {
 		return err
 	}
-	return v.Tx.DeepSet(address.Bytes(), accBytes, AccountsCfg)
+	v.Tx.Lock()
+	defer v.Tx.Unlock()
+	if err := v.Tx.DeepSet(accountAddress.Bytes(), accBytes, AccountsCfg); err != nil {
+		return err
+	}
+	return v.Tx.DeepSet(txSender.Bytes(), senderBytes, AccountsCfg)
 }
 
-func (v *State) CreateAccount(address common.Address, infoURI string, delegates []common.Address, initBalance uint64) error {
+func (v *State) CreateAccount(accountAddress common.Address, infoURI string, delegates []common.Address, initBalance uint64, faucetSender common.Address, faucetNonce uint64) error {
 	// check valid address
-	if address.String() == types.EthereumZeroAddress {
+	if accountAddress.String() == types.EthereumZeroAddressString {
 		return fmt.Errorf("invalid address")
 	}
 	// check not created
-	acc, err := v.GetAccount(address, false)
+	acc, err := v.GetAccount(accountAddress, false)
 	if err != nil {
-		return fmt.Errorf("cannot create account %s: %v", address.String(), err)
+		return fmt.Errorf("cannot create account %s: %v", accountAddress.String(), err)
 	}
 	if acc != nil {
-		return fmt.Errorf("account %s already exists", address.String())
+		return fmt.Errorf("account %s already exists", accountAddress.String())
 	}
+	acc = &Account{}
 	// account not found, creating it
 	// check valid infoURI, must be set on creation
-	if infoURI == "" {
-		return fmt.Errorf("cannot set an empty URI")
-	}
-	newAccount := &Account{}
-	newAccount.InfoURI = infoURI
+	acc.InfoURI = infoURI
 	if len(delegates) > 0 {
-		newAccount.DelegateAddrs = make([][]byte, len(delegates))
-		for i := 0; i < len(delegates); i++ {
-			if delegates[i].String() != types.EthereumZeroAddress {
-				newAccount.DelegateAddrs = append(newAccount.DelegateAddrs, delegates[i].Bytes())
+		acc.DelegateAddrs = make([][]byte, len(delegates))
+		for _, v := range delegates {
+			if v.String() != types.EthereumZeroAddressString {
+				acc.DelegateAddrs = append(acc.DelegateAddrs, v.Bytes())
 			}
 		}
 	}
-	newAccount.Balance = initBalance
-	v.Tx.Lock()
-	defer v.Tx.Unlock()
-	accBytes, err := newAccount.Marshal()
-	if err != nil {
-		return err
+	acc.Balance = initBalance
+
+	// handle faucet claim
+	if !bytes.Equal(faucetSender.Bytes(), types.EthereumZeroAddressBytes[:]) {
```

Comment:
I think faucet address should be a pointer and check if its nil or not. But in any case I feel like this faucet stuff should not be here. Once thing is create the Account and the other is fund it with faucet payload.